asab/library/providers/git.py

```diff
@@ -0,0 +1,145 @@
+import os
+import tempfile
+import logging
+import hashlib
+
+import pygit2
+
+from .filesystem import FileSystemLibraryProvider
+from ...config import Config
+
+#
+
+L = logging.getLogger(__name__)
+
+#
+
+
+class GitLibraryProvider(FileSystemLibraryProvider):
+	"""
+	Read-only git provider to read from remote repository.
+	It clones a remote git repository to a temporary directory and then uses the
+	FileSystemLibraryProvider to read the files.
+	To read from local git repository, please use FileSystemProvider.
+
+	Configuration:
+	(Use either deploytoken, publickey+privatekey for SSH option, or username and password and HTTP access.)
+
+
+	```
+	[library]
+	providers=git+<URL or deploy token>
+
+	[library:git]
+	publickey=<absolute path to file>
+	privatekey=<absolute path to file>
+	username=johnsmith
+	password=secretpassword
+	repodir=<optional location of the repository cache>
+	```
+	"""
+	def __init__(self, library, path):
+		self.LastCommit = None
+		self.URL = path[4:]
+
+		self.Callbacks = pygit2.RemoteCallbacks(get_git_credentials(self.URL))
+
+		# TODO: Check `repodir`
+		tempdir = tempfile.gettempdir()
+		self.RepoPath = os.path.join(
+			tempdir,
+			"asab.library.git",
+			hashlib.sha256(self.URL.encode('utf-8')).hexdigest()
+		)
+
+		try:
```

Comment:
These are time-consuming bits, delayed library inicialization is needed here.

Reply:
Yes, but library service initialization is waiting for all the providers to get ready. I cannot make cloning and fetching any faster...

Reply:
Sure, but you can make it "asynchronous", that's why we have `await self._set_ready()` ... check Zookeeper provider, it has long init time as well.